File: data/src/main/java/org/cryptomator/data/cloud/onedrive/OnedriveCloudContentRepository.java
package org.cryptomator.data.cloud.onedrive;

import android.content.Context;

import com.microsoft.graph.core.GraphErrorCodes;

import org.cryptomator.data.cloud.InterceptingCloudContentRepository;
import org.cryptomator.data.cloud.onedrive.graph.ClientException;
import org.cryptomator.domain.CloudNode;
import org.cryptomator.domain.OnedriveCloud;
import org.cryptomator.domain.exception.BackendException;
import org.cryptomator.domain.exception.FatalBackendException;
import org.cryptomator.domain.exception.NetworkConnectionException;
import org.cryptomator.domain.exception.NoSuchCloudFileException;
import org.cryptomator.domain.exception.authentication.WrongCredentialsException;
import org.cryptomator.domain.repository.CloudContentRepository;
import org.cryptomator.domain.usecases.ProgressAware;
import org.cryptomator.domain.usecases.cloud.DataSource;
import org.cryptomator.domain.usecases.cloud.DownloadState;
import org.cryptomator.domain.usecases.cloud.UploadState;
import org.cryptomator.util.Optional;

import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.net.SocketTimeoutException;
import java.util.List;

import static org.cryptomator.util.ExceptionUtil.contains;

class OnedriveCloudContentRepository extends InterceptingCloudContentRepository<OnedriveCloud, OnedriveNode, OnedriveFolder, OnedriveFile> {

	private final OnedriveCloud cloud;

	public OnedriveCloudContentRepository(OnedriveCloud cloud, Context context) {
		super(new Intercepted(cloud, context));
		this.cloud = cloud;
	}

	@Override
	protected void throwWrappedIfRequired(Exception e) throws BackendException {
		throwNetworkConnectionExceptionIfRequired(e);
		throwWrongCredentialsExceptionIfRequired(e);
	}

	private void throwNetworkConnectionExceptionIfRequired(Exception e) throws NetworkConnectionException {
		if (contains(e, SocketTimeoutException.class)) {
			throw new NetworkConnectionException(e);
		}
	}

	private void throwWrongCredentialsExceptionIfRequired(Exception e) {
		if (isAuthenticationError(e)) {
			throw new WrongCredentialsException(cloud);
		}
	}

	private boolean isAuthenticationError(Throwable e) {
		return e != null //
				&& ((e instanceof ClientException && ((ClientException) e).errorCode().equals(GraphErrorCodes.AUTHENTICATION_FAILURE)) //
						|| isAuthenticationError(e.getCause()));
	}

	private static class Intercepted implements CloudContentRepository<OnedriveCloud, OnedriveNode, OnedriveFolder, OnedriveFile> {

		private final OnedriveImpl oneDriveImpl;

		public Intercepted(OnedriveCloud cloud, Context context) {
			this.oneDriveImpl = new OnedriveImpl(cloud, context, new OnedriveIdCache());
		}

		@Override
		public OnedriveFolder root(OnedriveCloud cloud) {
			return oneDriveImpl.root();
		}

		@Override
		public OnedriveFolder resolve(OnedriveCloud cloud, String path) {
			return oneDriveImpl.resolve(path);
		}

		@Override
		public OnedriveFile file(OnedriveFolder parent, String name) {
			return oneDriveImpl.file(parent, name);
		}

		@Override
		public OnedriveFile file(OnedriveFolder parent, String name, Optional<Long> size) {
			return oneDriveImpl.file(parent, name, size);
		}

		@Override
		public OnedriveFolder folder(OnedriveFolder parent, String name) {
			return oneDriveImpl.folder(parent, name);
		}

		@Override
		public boolean exists(OnedriveNode node) throws BackendException {
			return oneDriveImpl.exists(node);
		}

		@Override
		public List<CloudNode> list(OnedriveFolder folder) throws BackendException {
			return oneDriveImpl.list(folder);
		}

		@Override
		public OnedriveFolder create(OnedriveFolder folder) throws BackendException {
			return oneDriveImpl.create(folder);
		}

		@Override
		public OnedriveFolder move(OnedriveFolder source, OnedriveFolder target) throws BackendException {
			return (OnedriveFolder) oneDriveImpl.move(source, target);
		}

		@Override
		public OnedriveFile move(OnedriveFile source, OnedriveFile target) throws BackendException {
			return (OnedriveFile) oneDriveImpl.move(source, target);
		}

		@Override
		public OnedriveFile write(OnedriveFile file, DataSource data, ProgressAware<UploadState> progressAware, boolean replace, long size) throws BackendException {
			try {
				return oneDriveImpl.write(file, data, progressAware, replace, size);
			} catch (BackendException e) {
				if (contains(e, NoSuchCloudFileException.class)) {
					throw new NoSuchCloudFileException(file.getName());
				}
				throw e;
			}
		}

		@Override
		public void read(OnedriveFile file, Optional<File> tmpEncryptedFile, OutputStream data, ProgressAware<DownloadState> progressAware) throws BackendException {
			try {
				oneDriveImpl.read(file, tmpEncryptedFile, data, progressAware);
			} catch (IOException | BackendException e) {
				if (contains(e, NoSuchCloudFileException.class)) {
					throw new NoSuchCloudFileException(file.getName());
				} else if (e instanceof IOException) {
					throw new FatalBackendException(e);
				} else if (e instanceof BackendException) {
					throw (BackendException) e;
				}
			}
		}

		@Override
		public void delete(OnedriveNode node) throws BackendException {
			oneDriveImpl.delete(node);
		}

		@Override
		public String checkAuthenticationAndRetrieveCurrentAccount(OnedriveCloud cloud) throws BackendException {
			return oneDriveImpl.currentAccount();
		}

		@Override
		public void logout(OnedriveCloud cloud) {
			oneDriveImpl.logout();
		}
	}

}

File: data/src/main/java/org/cryptomator/data/cloud/onedrive/OnedriveCloudContentRepository.kt
package org.cryptomator.data.cloud.onedrive

import android.content.Context
import com.microsoft.graph.core.GraphErrorCodes
import com.microsoft.graph.http.GraphServiceException
import com.microsoft.graph.requests.GraphServiceClient
import com.microsoft.identity.common.java.exception.ClientException
import org.cryptomator.data.cloud.InterceptingCloudContentRepository
import org.cryptomator.domain.OnedriveCloud
import org.cryptomator.domain.exception.BackendException
import org.cryptomator.domain.exception.FatalBackendException
import org.cryptomator.domain.exception.NetworkConnectionException
import org.cryptomator.domain.exception.NoSuchCloudFileException
import org.cryptomator.domain.exception.authentication.WrongCredentialsException
import org.cryptomator.domain.repository.CloudContentRepository
import org.cryptomator.domain.usecases.ProgressAware
import org.cryptomator.domain.usecases.cloud.DataSource
import org.cryptomator.domain.usecases.cloud.DownloadState
import org.cryptomator.domain.usecases.cloud.UploadState
import org.cryptomator.util.ExceptionUtil
import java.io.File
import java.io.IOException
import java.io.OutputStream
import java.net.SocketTimeoutException
import okhttp3.Request

internal class OnedriveCloudContentRepository(private val cloud: OnedriveCloud, context: Context, graphServiceClient: GraphServiceClient<Request>)
	: InterceptingCloudContentRepository<OnedriveCloud, OnedriveNode, OnedriveFolder, OnedriveFile>(Intercepted(cloud, context, graphServiceClient)) {

	@Throws(BackendException::class)
	override fun throwWrappedIfRequired(e: Exception) {
		throwNetworkConnectionExceptionIfRequired(e)
		throwWrongCredentialsExceptionIfRequired(e)
	}

	@Throws(NetworkConnectionException::class)
	private fun throwNetworkConnectionExceptionIfRequired(e: Exception) {
		if (ExceptionUtil.contains(e, SocketTimeoutException::class.java)) {
			throw NetworkConnectionException(e)
		}
	}

	private fun throwWrongCredentialsExceptionIfRequired(e: Exception) {
		if (isAuthenticationError(e)) {
			throw WrongCredentialsException(cloud)
		}
	}

	private fun isAuthenticationError(e: Throwable?): Boolean {
		return (e != null //
				&& (e is ClientException && e.errorCode == GraphErrorCodes.AUTHENTICATION_FAILURE.name //
				|| e is GraphServiceException && e.serviceError?.code?.equals("InvalidAuthenticationToken") == true
				|| isAuthenticationError(e.cause)))
	}

	private class Intercepted(cloud: OnedriveCloud, context: Context, graphServiceClient: GraphServiceClient<Request>) : CloudContentRepository<OnedriveCloud, OnedriveNode, OnedriveFolder, OnedriveFile> {

		private val oneDriveImpl: OnedriveImpl = OnedriveImpl(cloud, context, graphServiceClient, OnedriveIdCache())

		override fun root(cloud: OnedriveCloud): OnedriveFolder {
			return oneDriveImpl.root()
		}

		override fun resolve(cloud: OnedriveCloud, path: String): OnedriveFolder {
			return oneDriveImpl.resolve(path)
		}

		override fun file(parent: OnedriveFolder, name: String): OnedriveFile {
			return oneDriveImpl.file(parent, name)
		}

		override fun file(parent: OnedriveFolder, name: String, size: Long?): OnedriveFile {
			return oneDriveImpl.file(parent, name, size)
		}

		override fun folder(parent: OnedriveFolder, name: String): OnedriveFolder {
			return oneDriveImpl.folder(parent, name)
		}

		@Throws(BackendException::class)
		override fun exists(node: OnedriveNode): Boolean {
			return oneDriveImpl.exists(node)
		}

		@Throws(BackendException::class)
		override fun list(folder: OnedriveFolder): List<OnedriveNode> {
			return oneDriveImpl.list(folder)
		}

		@Throws(BackendException::class)
		override fun create(folder: OnedriveFolder): OnedriveFolder {
			return oneDriveImpl.create(folder)
		}

		@Throws(BackendException::class)
		override fun move(source: OnedriveFolder, target: OnedriveFolder): OnedriveFolder {
			return oneDriveImpl.move(source, target) as OnedriveFolder
		}

		@Throws(BackendException::class)
		override fun move(source: OnedriveFile, target: OnedriveFile): OnedriveFile {
			return oneDriveImpl.move(source, target) as OnedriveFile
		}

		@Throws(BackendException::class)
		override fun write(file: OnedriveFile, data: DataSource, progressAware: ProgressAware<UploadState>, replace: Boolean, size: Long): OnedriveFile {
			return try {
				oneDriveImpl.write(file, data, progressAware, replace, size)
			} catch (e: BackendException) {
				if (ExceptionUtil.contains(e, NoSuchCloudFileException::class.java)) {
					throw NoSuchCloudFileException(file.name)
				}
				throw e
			}
		}

		@Throws(BackendException::class)
		override fun read(file: OnedriveFile, encryptedTmpFile: File?, data: OutputStream, progressAware: ProgressAware<DownloadState>) {
			try {
				oneDriveImpl.read(file, encryptedTmpFile, data, progressAware)
			} catch (e: IOException) {
				when {
					ExceptionUtil.contains(e, NoSuchCloudFileException::class.java) -> {
						throw NoSuchCloudFileException(file.name)
					}
					else -> {
						throw FatalBackendException(e)
					}
				}
			} catch (e: BackendException) {
				when {
					ExceptionUtil.contains(e, NoSuchCloudFileException::class.java) -> {
						throw NoSuchCloudFileException(file.name)
					}
					else -> {
						throw e
					}
				}
			}
		}

		@Throws(BackendException::class)
		override fun delete(node: OnedriveNode) {
			oneDriveImpl.delete(node)
		}

		@Throws(BackendException::class)
		override fun checkAuthenticationAndRetrieveCurrentAccount(cloud: OnedriveCloud): String {
			return oneDriveImpl.currentAccount(cloud.username())
		}

		override fun logout(cloud: OnedriveCloud) {
			oneDriveImpl.logout()
		}

	}
}
